### PR TITLE
fix: only allow promoting successful template versions

### LINKED ISSUE #9904: do not allow promotion of a failed template version
This can lead to "create workspace" failing with an error about the template version failing

### PR DESCRIPTION
**Pull request by aslilac**
Closes #9904

- Add a check on the backend to prevent promoting an invalid version which has a failed/canceled/unfinished build
- Disable the "promote" button on the frontend to match
- Add "failed", "canceling", "building", and "pending" badges in the UI for versions with corresponding states
- Avoid showing the "newest" badge on versions that aren't valid
- ~~Hide "canceled" versions from the UI entirely~~

### CODE REVIEW

**Comment by aslilac** [comment]
sorry the diff is noisy, I split the renaming out into https://github.com/coder/coder/pull/10003

File: site/src/pages/TemplatePage/TemplateVersionsPage/VersionRow.tsx [hunk 1]
```diff
@@ -32,6 +32,11 @@ export const VersionRow: React.FC<VersionRowProps> = ({
     onClick: () => navigate(version.name),
   });
 
+  const jobStatus = version.job.status;
+  if (jobStatus === "canceled") {
+    return null;
+  }
```
Comment (BrunoQuaresma) [inline site/src/pages/TemplatePage/TemplateVersionsPage/VersionRow.tsx:38]: I feel this filtering should be happening in the parent and not here 🤔

**Review by BrunoQuaresma — COMMENTED** [review]
FE code looks good but since most of the changes are related to Go, I would also request a review from someone with more expertise on that

File: site/src/pages/TemplatePage/TemplateVersionsPage/VersionRow.tsx (hunk 1 above)
Comment (aslilac) [reply site/src/pages/TemplatePage/TemplateVersionsPage/VersionRow.tsx:38]: after thinking about this a little more, I'm realizing this probably shouldn't be the frontends job at all. messes with pagination, etc. if we wanna filter them it should be on the backend.

File: coderd/templateversions.go [hunk 2]
```diff
@@ -1040,6 +1041,22 @@ func (api *API) patchActiveTemplateVersion(rw http.ResponseWriter, r *http.Reque
...
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error fetching template version job status.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	jobStatus := db2sdk.ProvisionerJobStatus(job)
+	if jobStatus != codersdk.ProvisionerJobSucceeded {
+		httpapi.Write(ctx, rw, http.StatusPreconditionFailed, codersdk.Response{
```
Comment (spikecurtis) [inline coderd/templateversions.go:1054]: This should be 403 Forbidden.

PreconditionFailed refers to [preconditions](https://httpwg.org/specs/rfc9110.html#conditional.requests) that the HTTP client attached to the request in the headers.  We do not use these on our API and so should not return this status.

Note that 403 Forbidden is not just for insufficient permission, it should be used any time the request is well-formed and authentic, but we refuse to process.

Comment (aslilac) [reply coderd/templateversions.go]: ha, I saw this code used in one other place that seems similar, but after searching it looks like that was the only usage of it in the whole codebase. :p swapped them all to forbidden.

**Review by spikecurtis — APPROVED** [review]

**Pull request merged**